engine/src/rocksdb_engine.rs

```diff
@@ -60,7 +72,413 @@ impl RocksEngine {
+    snap_files: Vec<SnapFile>,
+    /// current file index
+    file_index: usize,
+    /// current file
+    current_file: Option<File>,
+    /// current file pos
+    current_file_pos: u64,
+}
+
+impl RocksSnapshot {
+    /// New empty `RocksSnapshot`
+    fn new<P>(dir: P) -> Result<Self, EngineError>
+    where
+        P: Into<PathBuf>,
+    {
+        let dir = dir.into();
+        if !dir.exists() {
+            fs::create_dir_all(&dir)?;
+        }
+        Ok(RocksSnapshot {
+            meta: Meta::new(),
+            dir,
+            snap_files: Vec::new(),
+            file_index: 0,
+            current_file: None,
+            current_file_pos: 0,
+        })
+    }
+
+    /// Create a new snapshot for sending
+    fn new_for_sending<P>(dir: P) -> Result<RocksSnapshot, EngineError>
+    where
+        P: Into<PathBuf>,
+    {
+        let dir = dir.into();
+        let files = fs::read_dir(&dir)?;
+        let mut s = Self::new(dir)?;
+        for file in files {
+            let entry = file?;
+            let filename = entry.file_name().into_string().map_err(|_e| {
+                EngineError::InvalidArgument("cannot convert filename to string".to_owned())
+            })?;
+            let size = entry.metadata()?.len();
+            s.snap_files.push(SnapFile {
+                filename: filename.clone(),
+                written_size: 0,
+                size,
+            });
+        }
+        s.gen_snap_meta()?;
+
+        Ok(s)
+    }
+
+    /// Apply snapshot meta
+    fn apply_snap_meta(&mut self, meta: SnapMeta) {
+        self.snap_files = meta
+            .files
+            .into_iter()
+            .map(|(filename, size)| SnapFile {
+                filename,
+                size,
+                ..Default::default()
+            })
+            .collect::<Vec<_>>();
+    }
+
+    /// Generate snapshot meta
+    fn gen_snap_meta(&mut self) -> Result<(), EngineError> {
+        let files = SnapMeta {
+            files: self
+                .snap_files
+                .iter()
+                .map(|sf| (sf.filename.clone(), sf.size))
+                .collect::<Vec<_>>(),
+        };
+        let meta_bytes = bincode::serialize(&files).map_err(|e| {
+            EngineError::UnderlyingError(format!("cannot serialize snapshot meta: {e}"))
+        })?;
+        let len = meta_bytes.len().numeric_cast::<u64>();
+        let mut data = Vec::new();
+        data.extend(&len.to_le_bytes());
+        data.extend(meta_bytes);
+        self.meta.data = Cursor::new(data);
+        Ok(())
+    }
+
+    /// path of current file
+    fn current_file_path(&self) -> PathBuf {
+        let Some(current_filename) = self.snap_files.get(self.file_index).map(|sf| &sf.filename) else {
+            unreachable!("this method must be called when self.file_index < self.snap_files.len()")
+        };
+        self.dir.join(current_filename)
+    }
+
+    /// tmp path of current file
+    fn current_file_tmp_path(&self) -> PathBuf {
```

Comment:
The `current_file_tmp_path` is very similar to the `current_file_path`. I think we'd better merge the together.
```suggestion
    fn current_file_path(&self, tmp: bool) -> PathBuf {
        let Some(current_filename) = self.snap_files.get(self.file_index).map(|sf| &sf.filename) else {
            unreachable!("this method must be called when self.file_index < self.snap_files.len()")
        };
        if tmp {
            self.dir.join(current_filename)
        } else {
            self.dir.join(format!("{current_filename}.tmp"))
        }
    }
```

Reply:
fixed